common/src/primitives/signed_amount.rs

```diff
@@ -0,0 +1,1039 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(clippy::eq_op)]
+
+use std::iter::Sum;
+
+use super::{amount::UnsignedIntType, Amount};
+
+pub type SignedIntType = i128;
+
+/// A signed fixed-point type for amounts used in accounting, specifically
+/// The smallest unit of count is called an atom
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SignedAmount {
+    val: SignedIntType,
+}
+
+impl SignedAmount {
+    pub const MAX: Self = Self::from_atoms(SignedIntType::MAX);
+    pub const MIN: Self = Self::from_atoms(SignedIntType::MIN);
+
+    pub const fn from_atoms(v: SignedIntType) -> Self {
+        SignedAmount { val: v }
+    }
+
+    pub fn into_atoms(&self) -> SignedIntType {
+        self.val
+    }
+
+    pub fn from_unsigned(amount: Amount) -> Option<Self> {
+        let unsigned_atoms = amount.into_atoms();
+        let atoms: SignedIntType = unsigned_atoms.try_into().ok()?;
+        Some(Self::from_atoms(atoms))
+    }
+
+    pub fn into_unsigned(self) -> Option<Amount> {
+        let atoms = self.val;
+        let unsigned_atoms: super::amount::UnsignedIntType = atoms.try_into().ok()?;
+        Some(Amount::from_atoms(unsigned_atoms))
+    }
+
+    pub fn into_fixedpoint_str(self, decimals: u8) -> String {
+        let negative = self.val < 0;
+
+        let abs_amount = Amount::from_atoms(self.val.unsigned_abs());
+
+        if negative {
+            "-".to_owned() + &abs_amount.into_fixedpoint_str(decimals)
+        } else {
+            abs_amount.into_fixedpoint_str(decimals)
+        }
+    }
+
+    pub fn from_fixedpoint_str(amount_str: &str, decimals: u8) -> Option<Self> {
+        let negative = amount_str.starts_with('-');
+        let amount_str = amount_str.strip_prefix('-').unwrap_or(amount_str);
+
+        let unsigned_amount = Amount::from_fixedpoint_str(amount_str, decimals)?;
+
+        // cover the corner case: SignedAmount::MIN
+        if negative && unsigned_amount.into_atoms() == UnsignedIntType::MAX / 2 + 1 {
+            return Some(SignedAmount {
+                val: SignedIntType::MIN,
+            });
+        }
+
+        let signed_amount = unsigned_amount.into_signed()?;
+        let signed_atoms = if negative {
+            -signed_amount.into_atoms()
+        } else {
+            signed_amount.into_atoms()
+        };
+        Some(Self::from_atoms(signed_atoms))
+    }
```

Comment:
Seems to me a fair portion of parsing and printing code could be reused between this and the unsigned amount

Reply:
Already done that using Amount. Not sure what you mean.

Reply:
Was looking at outdated version